sycl/doc/extensions/SubGroupBitmaskGroupFunctions/sub_group_bitmask_group_algorithms.asciidoc

```diff
@@ -0,0 +1,253 @@
+# Sub Group Bitmask Group Functions Extension for DPC++: SYCL_EXT_ONEAPI_SUB_GROUP_MASK_GROUP_FUNCTIONS
+:source-highlighter: coderay
+:coderay-linenums-mode: table
+:dpcpp: pass:[DPC++]
+
+// This section needs to be after the document title.
+:doctype: book
+:toc2:
+:toc: left
+:encoding: utf-8
+:lang: en
+
+:blank: pass:[ +]
+
+// Set the default source code type in this document to C++,
+// for syntax highlighting purposes.  This is needed because
+// docbook uses c++ and html5 uses cpp.
+:language: {basebackend@docbook:c++:cpp}
+
+
+== Notice
+
+Copyright (c) 2021-2021 Intel Corporation.  All rights reserved.
+
+IMPORTANT: This specification is a draft.
+
+NOTE: Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are
+trademarks of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc.
+used by permission by Khronos.
+
+This extension is written against the SYCL 2020 revision 3 specification.  All
+references below to the "core SYCL specification" or to section numbers in the
+SYCL specification refer to that revision.
+
+## Contributors
+
+* Gordon Brown
+* Tadej Ciglaric
+* Jack Kirk
+* Finlay Marno
+
+## Introduction
+
+This extension builds on the link:../SubGroupMask/SubGroupMask.asciidoc[`sub_group_mask`]
+extension to propose implementations of `group_barrier`, `joint_reduce`,
+`reduce_over_group` and the proposed `async_group_copy`, that perform operations using
+the work-items comprising a subset of a sub-group, using a bitmask to specify that subset.
+To aid the creation of helpful bitmasks, two member functions will be added to
+`nd_item`: `ext_oneapi_partition_sub_group` and `ext_oneapi_active_sub_group_items`. `ext_oneapi_partition_sub_group`
+will create a bitmask that splits the sub-group evenly. `ext_oneapi_active_sub_group_items`
+will create a bitmask representing the currently active work-items in the sub-group.
+
+This document also proposes that `async_work_group_copy` be deprecated and
+replaced with `async_group_copy`. `async_group_copy` is a non-member function,
+in line with the other group functions, which generalizes `async_work_group_copy`
+to also work with sub-groups.
```

Comment:
This feels a little out of place, to me.  Deprecating `async_work_group_copy` in favor of a general `async_group_copy` is a great idea, but it's not obvious that availability of the functionality should be tied to support for these new masked algorithms.

I think that the new `async_group_copy` function should be split out into its own extension.  This extension would then only introduce the overloads accepting a `sycl::sub_group_mask` parameter.

Reply:
Sounds like a good idea, I will do that.

Reply:
@Pennycook I've split that up. Should I open a new PR with just the `async_group_copy` stuff?

Reply:
Yes, I think so.  It'll be easier to review if it's in a separate PR.  Thanks!

Reply:
New PR opened at #4950